tool/src/org/antlr/v4/codegen/model/Choice.java

```diff
@@ -46,9 +49,21 @@ public void addPreambleOp(SrcOp op) {
 	}
 
 	public List<String[]> getAltLookaheadAsStringLists(IntervalSet[] altLookSets) {
-		List<String[]> altLook = new ArrayList<String[]>();
+		List<String[]> altLook = new ArrayList<>();
+		Target target = factory.getGenerator().getTarget();
+		Grammar grammar = factory.getGrammar();
 		for (IntervalSet s : altLookSets) {
-			altLook.add(factory.getGenerator().getTarget().getTokenTypesAsTargetLabels(factory.getGrammar(), s.toArray()));
+			if (target.supportsConstants()) {
```

Comment:
not sure what supportsConstants means.

Reply:
the "old way with lots of this|that|other"?

Reply:
It means if runtime is able to fold constants during compilation. As I wrote in https://github.com/antlr/antlr4/issues/3698, Python calculates the following expression:

```
(1 << C1) | (1 << C2) | (1 << C3) | (1 << C4) | (1 << C5) | (1 << C6) | \
        (1 << C7) | (1 << C8) | (1 << C9) | (1 << C10) | (1 << C11) | (1 << C12) | \
        (1 << C13) | (1 << C14) | (1 << C15) | (1 << C16)
```

every time when Python executes it because it's dynamic languages and it does not have the constant concept, i.e. `supportsConstans` returns false. C#, Java and other compilable languages are able to fold constant to `0xFFFF` on the compilation step and `supportsConstants` return true for them.

Reply:
The current code propagates all constants during generation if `supportsConstants` returns `true`:

```python
def r(self):
    localctx = TestParser.RContext(self, self._ctx, self.state)
    self.enterRule(localctx, 0, 0)
    try:
        self.enterOuterAlt(localctx, 1)
        self.state = 2
        self.match(1)
    except RecognitionException as re:
        localctx.exception = re
        self._errHandler.reportError(self, re)
        self._errHandler.recover(self, re)
    finally:
        self.exitRule()
    return localctx
```

instead of:

```python
def r(self):
    localctx = TestParser.RContext(self, self._ctx, self.state)
    self.enterRule(localctx, 0, self.RULE_r)
    try:
        self.enterOuterAlt(localctx, 1)
        self.state = 2
        self.match(TestParser.T__0)
    except RecognitionException as re:
        localctx.exception = re
        self._errHandler.reportError(self, re)
        self._errHandler.recover(self, re)
    finally:
        self.exitRule()
    return localctx
```

Reply:
Ah. Well, then it should have critical term "folding" haha.  How about `targetFoldsConstants` to indicate the target compiler or interpreter will handle it whereas python and friends need ANTLR to do it?  Or, change it to `foldConstants` to indicate ANTLR does it and add a longer comment on it? :D

Reply:
Not sure it's correct naming. Python supports constant folding, i.e. the following expression

```
(1 << 1) | (1 << 2) | (1 << 3) | (1 << 4) | (1 << 5) | (1 << 6) | \
        (1 << 7) | (1 << 8) | (1 << 9) | (1 << 10) | (1 << 11) | (1 << 12) | \
        (1 << 13) | (1 << 14) | (1 << 15) | (1 << 16)
```

is being folded to constant `0xFFFF`. But Python doesn't support constant propagation since Python is unable to determine whether a variable is constant or not.

Reply:
Oh! I get it now. Thanks.  I would suggest `targetFoldsConstantSymbols` or some such then to be explicit?

Reply:
`supportsConstantsPropagation`? target is excess word since the function is already a member of target.

Reply:
Propagation isn't commonly known by people.  I'm ok with that name but please explain in comment it means replacing constant symbols with literals.